Sources/NIOWebSocket/NIOWebSocketClientUpgrader.swift

```diff
@@ -0,0 +1,88 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+/// A `HTTPClientProtocolUpgrader` that knows how to do the WebSocket upgrade dance.
+///
+/// This upgrader assumes that the `HTTPClientUpgradeHandler` will create and send the upgrade request. 
+/// This upgrader also assumes that the `HTTPClientUpgradeHandler` will appropriately mutate the
+/// pipeline to remove the HTTP `ChannelHandler`s.
+public final class NIOWebClientSocketUpgrader: NIOHTTPClientProtocolUpgrader {
+    
+    /// RFC 6455 specs this as the required entry in the Upgrade header.
+    public var supportedProtocol: String = "websocket"
+    /// None of the websocket headers are actually defined as 'required'.
+    public var requiredUpgradeHeaders: [String] = []
+    
+    private let requestKey: String
+    private let maxFrameSize: Int
+    private let automaticErrorHandling: Bool
+    private let upgradePipelineHandler: (Channel, HTTPResponseHead) -> EventLoopFuture<Void>
+    
+    public init(requestKey: String,
+                maxFrameSize: Int = 1 << 14,
+                automaticErrorHandling: Bool = true,
+                upgradePipelineHandler: @escaping (Channel, HTTPResponseHead) -> EventLoopFuture<Void>) {
+
+        precondition(requestKey != "", "The request key must contain a valid Sec-WebSocket-Key")
+        precondition(maxFrameSize <= UInt32.max, "invalid overlarge max frame size")
+        self.requestKey = requestKey
+        self.upgradePipelineHandler = upgradePipelineHandler
+        self.maxFrameSize = maxFrameSize
+        self.automaticErrorHandling = automaticErrorHandling
+    }
+
+    public func addCustom(upgradeRequestHeaders: inout HTTPHeaders) {
```

Comment:
this function doesn't seem to be used, do we need it? If yes, we should give it a better name, document it, and also test it.

Reply:
They are part of the conformance to NIOHTTPClientProtocolUpgrader. It is tested as part of the end to end tests. 'testSimpleUpgradeSucceeds' reads the request coming out of the client and checks for these headers. I added some comments as to what it is doing. I can rename it on the protocol upgrader if you think that will make it clearer?

Reply:
Thanks, comment is good enough